<a href="https://colab.research.google.com/github/nataliaespector/CEIA_TPs_Procesamiento_Lenguaje_Natural_1/blob/main/Espector_PLN_Desafio_1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [143]:
%pip install numpy scikit-learn

## Importar librerías necesarias

In [144]:
from sklearn.feature_extraction.text import CountVectorizer, TfidfVectorizer
from sklearn.metrics.pairwise import cosine_similarity
from sklearn.naive_bayes import MultinomialNB, ComplementNB
from sklearn.metrics import f1_score, confusion_matrix, ConfusionMatrixDisplay

import matplotlib.pyplot as plt
import seaborn as sns

from sklearn.datasets import fetch_20newsgroups
import numpy as np

# 1. Vectorizar documentos
Tomar 5 documentos al azar y medir similaridad con el resto de los documentos. Estudiar los 5 documentos más similares de cada uno analizar si tiene sentido la similaridad según el contenido del texto y la etiqueta de clasificación.

Carga de datos:

In [145]:
newsgroups_train = fetch_20newsgroups(subset='train', remove=('headers', 'footers', 'quotes'))
newsgroups_test = fetch_20newsgroups(subset='test', remove=('headers', 'footers', 'quotes'))

Se seleccionan y muestran 5 documentos al azar (índices 5, 10, 16, 21, 26), con las categorías a las que pertenecen:

In [146]:
index = [5,10,16,21,26]
for idx in index:
  clase = newsgroups_train.target_names[newsgroups_train.target[idx]]
  print(f"=== Texto {idx} - Categoría {clase}===")
  print(newsgroups_train.data[idx])
  print("\n" + "=" * 80 + "\n")

=== Texto 5 - Categoría talk.politics.guns===





Of course.  The term must be rigidly defined in any bill.


I doubt she uses this term for that.  You are using a quote allegedly
from her, can you back it up?




I read the article as presenting first an argument about weapons of mass
destruction (as commonly understood) and then switching to other topics.
The first point evidently was to show that not all weapons should be
allowed, and then the later analysis was, given this understanding, to
consider another class.






=== Texto 10 - Categoría rec.motorcycles===
I have a line on a Ducati 900GTS 1978 model with 17k on the clock.  Runs
very well, paint is the bronze/brown/orange faded out, leaks a bit of oil
and pops out of 1st with hard accel.  The shop will fix trans and oil 
leak.  They sold the bike to the 1 and only owner.  They want $3495, and
I am thinking more like $3K.  Any opinions out there?  Please email me.
Thanks.  It would be a nice stable mate to the Beemer.  Then I

Instanciar el vectorizador:

In [147]:
tfidfvect = TfidfVectorizer()

Obtener X_train e y_train vectorizados:

In [148]:
X_train = tfidfvect.fit_transform(newsgroups_train.data)
y_train = newsgroups_train.target

Medir similaridad coseno de cada documento con todos los documentos de train y analizar 5 más similares:

Texto 5: El texto pertenecía a la clase política/armas, y trata sobre armas de destrucción masiva. Observando los textos más similares, vemos que:
- 198 - Pertenece a la misma categoría y también habla sobre armas de destrucción masiva
- 7963 - Pertenece a una categoría completamente distinta, a religión/cristiana. Analizando el texto, si bien es un texto religioso, habla sobre muerte, oscuridad, suicidio, peligro, que son palabras que pueden encontrarse en el mismo contexto que las armas.
- 1292 - Pertenece a la categoría política/mideast. Tiene sentido la similaridad, ya que habla de un contexto de guerra, y menciona términos como violación de derechos humanos, ataque, muerte, cadáver, soldados, enemigos, guerra.
- 6894 - Pertenece a la misma categoría. Tiene sentido, ya que habla de muerte, armas, munición, violencia, rehenes, tragedia, asesinato, militares.
- 1533 - Pertenece a la categoría religión/cristiana, al igual que el texto 7963. Sin embargo, este es un texto que habla sobre una relación romántica, por lo que cuesta encontrar la similaridad con los otros. Podría ser debido a que en algunas partes del texto habla sobre temas médicos (escuela de medicina, beneficios médicos), pero no está tan claro como en los otros textos.

In [149]:
idx=5
cossim = cosine_similarity(X_train[idx], X_train)[0]

In [150]:
mostsim = np.argsort(cossim)[::-1][1:6]

In [151]:
for i in mostsim:
  clase = newsgroups_train.target_names[y_train[i]]
  print(f"=== Texto {i} - Categoría {clase}===")
  print(newsgroups_train.data[i])
  print("\n" + "=" * 80 + "\n")

=== Texto 198 - Categoría talk.politics.guns===





I don't sign any blank checks.

When Doug Foxvog says "weapons of mass destruction," he means CBW and
nukes.  When Sarah Brady says "weapons of mass destruction" she means
Street Sweeper shotguns and semi-automatic SKS rifles.  When John
Lawrence Rutledge says "weapons of mass destruction," and then immediately
follows it with:


...what does Rutledge mean by the term?
-- 


=== Texto 7963 - Categoría soc.religion.christian===

[ -and many others mailed me.  Here is a reply to one of the letters.  Seems to
me that atheist do not like the doctrine of hell!]



Your using 20th century concepts to interprete 1st century writers.  Of course,
in your termonology, God could not "cease to exist."  However, that is not what
death ever means in the Scriptures.  If you will study the word, you will see
that it signifies "separation."  Death is separation, not ceastation.  This is
the reason for the agony of the cross.  For the first time in et

Texto 10: El texto pertenecía a la clase motocicletas, y hablaba sobre distintos aspectos de la moto y el motor. Observando los textos más similares, vemos que en todos los casos los textos pertenecen a la categoría motocicletas, y además:
- 3543 y 9171: Están firmados por la misma persona, por lo que coincide en una gran cantidad de palabras (nombre, mail, frase que lo identifica, etc).
- 4930, 5063, 4211: tocan los mismos temas, e incluyen muchas veces palabras como motorcycle, motor, ride, bike, rider, oil, DoD.


In [152]:
idx=10
cossim = cosine_similarity(X_train[idx], X_train)[0]

In [153]:
mostsim = np.argsort(cossim)[::-1][1:6]

In [154]:
for i in mostsim:
  clase = newsgroups_train.target_names[y_train[i]]
  print(f"=== Texto {i} - Categoría {clase}===")
  print(newsgroups_train.data[i])
  print("\n" + "=" * 80 + "\n")

=== Texto 3543 - Categoría rec.motorcycles===

Now you know why I am just a DOD member.  I like bikes and clubs but
the politics and other b*llsh*t is a real turn-off.
-- 
-----------------------------------------------------------------------
"Tuba" (Irwin)      "I honk therefore I am"     CompuTrac-Richardson,Tx
irwin@cmptrc.lonestar.org    DoD #0826          (R75/6)


=== Texto 9171 - Categoría rec.motorcycles===

More like those who use their backs instead of their minds to make
their living who are usually ignorant and intolerant of anything outside
of their group or level of understanding.
-- 
-------------------------------------------------------------------------
"Tuba" (Irwin)      "I honk therefore I am"       CompuTrac-Richardson,Tx
irwin@cmptrc.lonestar.org    DoD #0826            (R75/6)


=== Texto 4930 - Categoría rec.motorcycles===
This is a periodic posting intended to answer the Frequently Asked
Question: What is the DoD? It is posted the first of each month, with
an

Texto 16: El texto pertenecía a la clase computación/gráficos, y hablaba sobre imágenes en formato TIFF. Observando los textos más similares, vemos que en todos los casos los textos pertenecen a la misma categoría, y en todos los casos tratan sobre imágenes TIFF, por lo que en este caso la similaridad está clara. Todos incluyen palabras como TIFF, programs, image, file, filename, format.


In [155]:
idx=16
cossim = cosine_similarity(X_train[idx], X_train)[0]

In [156]:
mostsim = np.argsort(cossim)[::-1][1:6]

In [157]:
for i in mostsim:
  clase = newsgroups_train.target_names[y_train[i]]
  print(f"=== Texto {i} - Categoría {clase}===")
  print(newsgroups_train.data[i])
  print("\n" + "=" * 80 + "\n")

=== Texto 2810 - Categoría comp.graphics===

I'm sure it is, and I am not amused.  Every time I read that part of the
TIFF spec, it infuriates me- and I'm none too happy about the
complexity of the spec anyway- because I think their "arbitrary but
carefully chosen number" is neither.  Additionally, I find their
choice of 4 bytes to begin a file with meaningless of themselves- why
not just use the letters "TIFF"?

(And no, I don't think they should have bothered to support both word
orders either- and I've found that many TIFF readers actually
don't.)


=== Texto 1803 - Categoría comp.graphics===
Are there any TIFF to anything programs out there for the IBM?
Our scanner works into TIFF, and I can view it on CSHOW 8.1, but 
all of my other programs read errors.  Are there any basic Tiff to 
JPEG, GIF, PCX, BMP, etc...?

Thanks for the time...Email or post acceptable.
Joshuaf


=== Texto 3046 - Categoría comp.graphics===

I've got the 6.0 spec (obviously since I quoted it in my last posti

Texto 21: El texto pertenecía a la clase deportes/hockey, y hablaba sobre distintos jugadores, en particular capitanes y ventas (traded) a otros equipos. Observando los textos más similares, vemos que en todos los casos los textos pertenecen a la misma categoría, y en todos los casos tratan sobre capitanes y ventas, por lo que en este caso la similaridad está clara. Todos incluyen palabras como captain y traded, y nombres propios de jugadores y equipos que se repiten en muchos de los textos.


In [158]:
idx=21
cossim = cosine_similarity(X_train[idx], X_train)[0]

In [159]:
mostsim = np.argsort(cossim)[::-1][1:6]

In [160]:
for i in mostsim:
  clase = newsgroups_train.target_names[y_train[i]]
  print(f"=== Texto {i} - Categoría {clase}===")
  print(newsgroups_train.data[i])
  print("\n" + "=" * 80 + "\n")

=== Texto 2858 - Categoría rec.sport.hockey===
: >And, while we are on the subject, has a captain ever been traded, 
: >resigned, or been striped of his title during the season? Any other 
: >team captain trivia would be appreciated.
;
: Wasn't Ron Francis captain of the Whalers when he was traded to
: Pittsburgh?

  And Rick Tochett was the captain of the Flyers when traded to the Pens
recently...

Caleb

And let us not forget that the New Jersey Devils traded
captain Kirk Muller for Stephen Richer and Chorske


=== Texto 11046 - Categoría rec.sport.hockey===
-=> Quoting Cire Y. Trehguad to All <=-
 : >And, while we are on the subject, has a captain ever been traded, 
 : >resigned, or been striped of his title during the season? Any other 
 : >team captain trivia would be appreciated.
 CYT> ;
 CYT> : Wasn't Ron Francis captain of the Whalers when he was traded to
 CYT> : Pittsburgh?

 CYT> And Rick Tochett was the captain of the Flyers when traded to the Pens
 CYT> recently...

 CYT> 

Texto 26: El texto pertenecía a la clase ciencia/electrónica, y hablaba sobre fuentes de energía (fósil, nuclear, gasolina, gas). Observando los textos más similares, vemos que en todos los casos los textos pertenecen a la misma categoría, y en todos los casos tratan sobre el mismo tema, por lo que la similaridad está clara. Todos incluyen palabras como fossil, fuel, plant, nuclear, power, water, cooling, temperature, tower.
Una particularidad de este caso es que en el texto más similar encontrado (2759) el primer párrafo es exactamente igual al del texto original, y luego se modifica.

In [161]:
idx=26
cossim = cosine_similarity(X_train[idx], X_train)[0]

In [162]:
mostsim = np.argsort(cossim)[::-1][1:6]

In [163]:
for i in mostsim:
  clase = newsgroups_train.target_names[y_train[i]]
  print(f"=== Texto {i} - Categoría {clase}===")
  print(newsgroups_train.data[i])
  print("\n" + "=" * 80 + "\n")

=== Texto 2759 - Categoría sci.electronics===
# >       Actually, fossil fuel plants run hotter than the usual 
# >boiling-water reactor nuclear plants.  (There's a gripe in the industry
# >that nuclear power uses 1900 vintage steam technology).  So it's
# >more important in nuclear plants to get the cold end of the system
# >as cold as possible.  Hence big cooling towers.  

When the utility gave up on that Cinnci, OH plant (Zimmer?) and
announced they were going to convert it to a coal-fired scheme, the
turbines were already in place, and they were the low-temp type. So the
plan was: Install a SECOND set of high temp turbines, and feed the
low-temp ones with the output of the new ones.

Never saw anything more on this. Did they ever really build it?


=== Texto 3809 - Categoría sci.electronics===
Excerpts from netnews.sci.electronics: 16-Apr-93 Re: What do Nuclear
Site's .. by R_Tim_Coslet@cup.portal. 

Great Explaination, however you left off one detail, why do you always
see them a

# 2. Construir un modelo de clasificación por prototipos (tipo zero-shot).
Clasificar los documentos de un conjunto de test comparando cada uno con todos los de entrenamiento y asignar la clase al label del documento del conjunto de entrenamiento con mayor similaridad.

In [164]:
X_test = tfidfvect.transform(newsgroups_test.data)
y_test = newsgroups_test.target

In [165]:
y_pred = []
for x in X_test:
  cossim = cosine_similarity(x, X_train)[0]
  mostsim = np.argmax(cossim)
  y_pred.append(y_train[mostsim])

In [166]:
f1_zeroshot=f1_score(y_test, y_pred, average='macro')
print("El f1-score del modelo de clasificación tipo zero-shot es",f1_zeroshot)

El f1-score del modelo de clasificación tipo zero-shot es 0.5049911553681621


# 3. Entrenar modelos de clasificación Naive Bayes para maximizar el desempeño de clasificación (f1-score macro) en el conjunto de datos de test.
Considerar cambiar parámteros de instanciación del vectorizador y los modelos y probar modelos de Naïve Bayes Multinomial y ComplementNB.

Modelo Naive Bayes Multinomial sin modificar parámetros:

In [167]:
clf_MNB = MultinomialNB()
clf_MNB.fit(X_train, y_train)
y_pred = clf_MNB.predict(X_test)
f1_multinomialNB = f1_score(y_test, y_pred, average='macro')

In [168]:
print("El f1-score del modelo de clasificación Naive Bayes Multinomial es",f1_multinomialNB)

El f1-score del modelo de clasificación Naive Bayes Multinomial es 0.5854345727938506


Modelo Naive Bayes Multinomial cambiando parámetros del modelo (alfa) para mejorar el f1-score:

In [169]:
clf_MNB = MultinomialNB(alpha=0.1)
clf_MNB.fit(X_train, y_train)
y_pred = clf_MNB.predict(X_test)
f1_multinomialNB = f1_score(y_test, y_pred, average='macro')

In [170]:
print("El f1-score del modelo de clasificación Naive Bayes Multinomial es",f1_multinomialNB)

El f1-score del modelo de clasificación Naive Bayes Multinomial es 0.6564514103512165


Modelo Complement Naive Bayes sin modificar parámetros:

In [171]:
clf_CNB = ComplementNB()
clf_CNB.fit(X_train, y_train)
y_pred = clf_CNB.predict(X_test)
f1_complementNB = f1_score(y_test, y_pred, average='macro')

In [172]:
print("El f1-score del modelo de clasificación Complement Naive Bayes es",f1_complementNB)

El f1-score del modelo de clasificación Complement Naive Bayes es 0.692953349950875


Modelo Complement Naive Bayes modificando parámetros del modelo (alfa) para mejorar el f1-score:

In [173]:
clf_CNB = ComplementNB(alpha=0.25)
clf_CNB.fit(X_train, y_train)
y_pred = clf_CNB.predict(X_test)
f1_complementNB = f1_score(y_test, y_pred, average='macro')

In [174]:
print("El f1-score del modelo de clasificación Complement Naive Bayes es",f1_complementNB)

El f1-score del modelo de clasificación Complement Naive Bayes es 0.7000110030082339


Se modifican los parámetros de instanciación del vectorizador para analizar bigramas. Se prueban nuevamente los modelos Multinomial NB y Complement NB con los mejores parámetros encontrados anteriormente, y se analiza el f1-score:

In [175]:
tfidfvect = TfidfVectorizer(ngram_range=(1,2))
X_train = tfidfvect.fit_transform(newsgroups_train.data)
y_train = newsgroups_train.target
X_test = tfidfvect.transform(newsgroups_test.data)
y_test = newsgroups_test.target

In [176]:
clf_MNB = MultinomialNB(alpha=0.1)
clf_MNB.fit(X_train, y_train)
y_pred = clf_MNB.predict(X_test)
f1_multinomialNB = f1_score(y_test, y_pred, average='macro')

In [177]:
print("El f1-score del modelo de clasificación Naive Bayes Multinomial y vectorizador usando bigramas es",f1_multinomialNB)

El f1-score del modelo de clasificación Naive Bayes Multinomial y vectorizador usando bigramas es 0.6220682060555511


In [178]:
clf_CNB = ComplementNB(alpha=0.25)
clf_CNB.fit(X_train, y_train)
y_pred = clf_CNB.predict(X_test)
f1_complementNB = f1_score(y_test, y_pred, average='macro')

In [179]:
print("El f1-score del modelo de clasificación Complement Naive Bayes y vectorizador usando bigramas es",f1_complementNB)

El f1-score del modelo de clasificación Complement Naive Bayes y vectorizador usando bigramas es 0.7013920563858125


No se observan mejoras en el f1-score en el caso del modelo MultinomialNB, y el valor es prácticamente el mismo en el caso de Complement Naive Bayes. Se realiza la misma prueba, modificando esta vez la consideración de stop words:

In [180]:
tfidfvect = TfidfVectorizer(stop_words='english')
X_train = tfidfvect.fit_transform(newsgroups_train.data)
y_train = newsgroups_train.target
X_test = tfidfvect.transform(newsgroups_test.data)
y_test = newsgroups_test.target

In [181]:
clf_MNB = MultinomialNB(alpha=0.1)
clf_MNB.fit(X_train, y_train)
y_pred = clf_MNB.predict(X_test)
f1_multinomialNB = f1_score(y_test, y_pred, average='macro')

In [182]:
print("El f1-score del modelo de clasificación Naive Bayes Multinomial y vectorizador considerando stop-words es",f1_multinomialNB)

El f1-score del modelo de clasificación Naive Bayes Multinomial y vectorizador considerando stop-words es 0.6725863920128103


In [183]:
clf_CNB = ComplementNB(alpha=0.25)
clf_CNB.fit(X_train, y_train)
y_pred = clf_CNB.predict(X_test)
f1_complementNB = f1_score(y_test, y_pred, average='macro')

In [184]:
print("El f1-score del modelo de clasificación Complement Naive Bayes y vectorizador considerando stop-words es",f1_complementNB)

El f1-score del modelo de clasificación Complement Naive Bayes y vectorizador considerando stop-words es 0.6982134900176366


Modificando la consideración de stop words se obtiene un mejor f1-score en el caso del modelo Multinomial NB, y prácticamente no se observa diferencia en el caso de Complement NB. Se realiza la prueba de modificar min_df (frecuencia mínima), de manera de filtrar términos raros que aparecen en muy pocos documentos:

In [185]:
tfidfvect = TfidfVectorizer(stop_words='english', min_df=2)
X_train = tfidfvect.fit_transform(newsgroups_train.data)
y_train = newsgroups_train.target
X_test = tfidfvect.transform(newsgroups_test.data)
y_test = newsgroups_test.target

In [186]:
clf_MNB = MultinomialNB(alpha=0.1)
clf_MNB.fit(X_train, y_train)
y_pred = clf_MNB.predict(X_test)
f1_multinomialNB = f1_score(y_test, y_pred, average='macro')

In [187]:
print("El f1-score del modelo de clasificación Naive Bayes Multinomial y vectorizador considerando stop-words y min_df=2 es",f1_multinomialNB)

El f1-score del modelo de clasificación Naive Bayes Multinomial y vectorizador considerando stop-words y min_df=2 es 0.6797635512454454


In [188]:
clf_CNB = ComplementNB(alpha=0.25)
clf_CNB.fit(X_train, y_train)
y_pred = clf_CNB.predict(X_test)
f1_complementNB = f1_score(y_test, y_pred, average='macro')

In [189]:
print("El f1-score del modelo de clasificación Complement Naive Bayes y vectorizador considerando stop-words y min_df=2 es",f1_complementNB)

El f1-score del modelo de clasificación Complement Naive Bayes y vectorizador considerando stop-words y min_df=2 es 0.6945348674610105


Agregando min_df=2, se observa una muy leve mejoría del f1-score en el caso de Multinomial NB, y prácticamente no se observan diferencias en el caso de Complement NB.

**Conclusión: considerando todas las pruebas realizadas, los mejores resultados (f1-score = 0.701) se obtuvieron utilizando bigramas en el vectorizador (ngram_range=(1,2)) y el modelo Complement Naive Bayes con alfa=0.25.**

# 4. Transponer la matriz documento-término.
De esa manera se obtiene una matriz término-documento que puede ser interpretada como una colección de vectorización de palabras. Estudiar ahora similaridad entre palabras tomando 5 palabras y estudiando sus 5 más similares. **La elección de palabras no debe ser al azar para evitar la aparición de términos poco interpretables, elegirlas "manualmente".**

Para este ejercicio, se vuelve a vectorizar utilizando los valores por defecto y se transpone la matriz documento-término:

In [190]:
tfidfvect = TfidfVectorizer()
X_train = tfidfvect.fit_transform(newsgroups_train.data)
y_train = newsgroups_train.target
X_test = tfidfvect.transform(newsgroups_test.data)
y_test = newsgroups_test.target

In [191]:
X_train_T = X_train.T

Se obtiene diccionario que va de índices a términos:

In [192]:
idx2word = {v: k for k,v in tfidfvect.vocabulary_.items()}

Se seleccionan palabras que fueron observadas como clave en los textos analizados previamente:

In [193]:
palabras_elegidas = ['weapons', 'motorcycle', 'program', 'captain', 'nuclear']

In [194]:
indices_elegidos = [tfidfvect.vocabulary_[word] for word in palabras_elegidas]

Se calcula la similaridad entre las palabras elegidas y el resto de las palabras utilizando la similaridad coseno:

In [195]:
word_similarity = cosine_similarity(X_train_T[indices_elegidos], X_train_T)

Obtener las 5 palabras más similares a cada una de las palabras elegidas:

In [196]:
for i, word_index in enumerate(indices_elegidos):
    similarities = word_similarity[i]

    most_similar_indices = np.argsort(similarities)[::-1][1:6]

    print(f"Palabras más similares a '{palabras_elegidas[i]}':")
    for idx in most_similar_indices:
        print(f"- {idx2word[idx]}")
    print("-" * 40)

Palabras más similares a 'weapons':
- sarin
- weapon
- millitary
- facist
- azerbadjan
----------------------------------------
Palabras más similares a 'motorcycle':
- wheelie
- shaft
- learner
- inquiries
- bluntly
----------------------------------------
Palabras más similares a 'program':
- allocate
- to
- windows
- the
- ftp
----------------------------------------
Palabras más similares a 'captain':
- striped
- traded
- resigned
- tochett
- chorske
----------------------------------------
Palabras más similares a 'nuclear':
- plants
- towers
- reactor
- fossil
- radioactives
----------------------------------------


En general, se observa que las palabras obtenidas como más similares en cada caso tienen sentido. En algunos casos, como 'weapons', 'program' y 'nuclear' las relaciones con las palabras están claras. En otros casos, la relación se ve más claramente si analizamos los textos que se estudiaron previamente.

Por ejemplo, en el caso de 'motorcycle', la palabra 'inquiries' en principio no tendría sentido, pero los textos analizados correspondían a preguntas/consultas que hacían los usuarios sobre las motos, por lo que allí podría estar la relación.

Otro ejemplo es el caso de 'captain', en donde vemos algunos nombres propios ('Tochett', 'Chorske'). Analizando los textos previos, vemos que esos nombres corresponden a capitanes de hockey. Si en los datos hay muchos documentos sobre hockey, entonces tiene sentido que los términos estén relacionados.